File: src/emailservice/client/email_service_stub.py
import os

from .client import marshal_request, unmarshal_response, send_request, DEFAULT_TIMEOUT, \
    EMAIL_SERVICE, SEND_ORDER_CONFIRMATION_RPC
from ..logger import getJSONLogger

logger = getJSONLogger('emailservice-stub')


def send_order_confirmation(request, headers=None):
    bin_req = marshal_request(request)
    resp_body, headers = send_request(greeter_addr, EMAIL_SERVICE, SEND_ORDER_CONFIRMATION_RPC, bin_req, headers,
                                      greeter_timeout)
    return unmarshal_response(resp_body, headers, SEND_ORDER_CONFIRMATION_RPC)


greeter_addr = os.getenv("EMAIL_SERVICE_ADDR")
if greeter_addr is None:
    logger.error("EMAIL_SERVICE_ADDR environment variable not set")
    raise EnvironmentError("EMAIL_SERVICE_ADDR environment variable not set")

t = os.getenv("EMAIL_SERVICE_TIMEOUT")
if t is None:
    greeter_timeout = DEFAULT_TIMEOUT
else:
    try:
        greeter_timeout = int(t)
        if greeter_timeout <= 0:
            greeter_timeout = DEFAULT_TIMEOUT
    except ValueError:
        greeter_timeout = DEFAULT_TIMEOUT

File: src/checkoutservice/client/email_service_stub.go
package client

import (
    "log"
    "net/http"
    "os"
    "strconv"

    pb "main/genproto"
)

var (
    emailServiceAddr    *string
    emailServiceTimeout *int
)

const (
    emailService             = "email-service"
    sendOrderConfirmationRPC = "send-order-confirmation"
)

// SendOrderConfirmation represents the EmailService/SendOrderConfirmation RPC.
// context can be sent as custom headers.
func SendOrderConfirmation(request *pb.SendOrderConfirmationRequest, header *http.Header) (*pb.Empty, error) {
    binReq, err := marshalRequest(request)
    if err != nil {
        return nil, err
    }

    respBody, header, err := sendRequest(*emailServiceAddr, emailService, sendOrderConfirmationRPC, &binReq, header, *emailServiceTimeout)
    if err != nil {
        return nil, err
    }

    msg, err := unmarshalResponse(respBody, header, sendOrderConfirmationRPC)
    if err != nil {
        return nil, err
    }

    return (*msg).(*pb.Empty), nil
}

// init loads the address and timeout variables.
func init() {
    a, ok := os.LookupEnv("EMAIL_SERVICE_ADDR")
    if !ok {
        log.Fatal("EMAIL_SERVICE_ADDR environment variable not set")
    }
    emailServiceAddr = &a

    t, ok := os.LookupEnv("EMAIL_SERVICE_TIMEOUT")
    if !ok {
        t := defaultTimeout
        emailServiceTimeout = &t
    } else {
        if t, err := strconv.Atoi(t); err != nil || t <= 0 {
            t = defaultTimeout
            emailServiceTimeout = &t
        } else {
            emailServiceTimeout = &t
        }
    }
}
